botocore/awsrequest.py

```diff
@@ -534,6 +557,84 @@ def reset_stream(self):
             raise UnseekableStreamError(stream_object=self.body)
 
 
+class AWSRequestCompressor:
+    """A class that can compress the body of an ``AWSRequest``."""
+
+    def compress(self, body, encodings, headers):
+        """Compresses the reqeust body using the specified encodings.
+
+        Set or append the `Content-Encoding` header with the matched encoding
+        if not present.
+        """
+        for encoding in encodings:
+            encoder = SUPPORTED_COMPRESSION_ENCODINGS.get(encoding)
+            if encoder is not None:
+                if 'Content-Encoding' not in headers:
+                    headers['Content-Encoding'] = encoding
+                elif encoding not in headers['Content-Encoding']:
+                    headers.replace_header(
+                        'Content-Encoding',
+                        f'{headers["Content-Encoding"]},{encoding}',
+                    )
+                return self._compress_body(body, encoder)
+            else:
+                logger.debug('Unsupported compression encoding: %s' % encoding)
+        return body
+
+    def _compress_body(self, body, encoder):
+        if isinstance(body, str):
+            return encoder(body.encode('utf-8'))
+        elif isinstance(body, (bytes, bytearray)):
+            return encoder(body)
+        elif hasattr(body, 'read'):
+            return encoder(b''.join(self._read_body(body)))
+        raise TypeError('Unable to compress body of type: %s' % type(body))
+
+    def _read_body(self, body):
+        while True:
+            # urllib3 and by extension the http stdlib uses a chunksize of 8192
+            # so we do the same here.
+            chunk = body.read(8192)
+            if not chunk:
+                break
+            if isinstance(chunk, str):
+                chunk = chunk.encode('utf-8')
+            yield chunk
+        if hasattr(body, 'seek'):
+            body.seek(0)
+
+    def should_compress_request(self, config, body, operation_model):
+        """Check if an AWSRequest object's body should be compressed."""
+        # compression must be explicitly disabled
+        if (
+            botocore.utils.ensure_boolean(config.disable_request_compression)
+            is not True
+        ):
+            min_size = config.request_min_compression_size_bytes
+            if min_size is None:
+                min_size = 10240
+            else:
+                min_size = int(min_size)
```

Comment:
I should have mentioned it in the PR description, but this was a very intentional part of the design. I chose not to use `ensure_boolean` and `int` in `conversion_func` because the former coerces any non-boolean value that isn't equal to `true` to `False` and the latter would provide a relatively opaque error message to the user. 

I made the tradeoff of being more closely aligned with the spec than how other config values are handled in botocore:
```
This setting is a boolean with values true or false (case-sensitive). The SDK MUST provide an error for any other value. The SDK MUST use false as the default value.
```

Reply:
To follow up on this: Can these checks and conversions actually be dropped even if we use `conversion_func`? That wouldn't cover when customers use their own `Config` object.

Reply:
We shouldn't be changing behavior from how all other options in the SDK work. If a non-True value is passed we default to False. We can go back and have the wording in the design adjusted but if there isn't a strong argument for this requirement, it's going to be an unexpected idiosyncrasy of this feature.

Reply:
Thanks Nate. I'll remove all of this custom code and just default to how other config settings are handled in the SDK.